src/browser/api_protocol/transport_strategy/elipc_strategy.ts

```diff
@@ -35,15 +35,21 @@ export class ElipcStrategy extends ApiTransportBase<MessagePackage> {
             if (strategyName !== this.constructor.name) {
                 next();
             } else if (typeof (action) === 'function') {
-                try {
-                    // singleFrameOnly check first so to prevent frame superceding when disabled.
-                    if (!data.singleFrameOnly === false || e.sender.isValidWithFrameConnect(e.frameRoutingId)) {
-                        action(identity, data, ack, nack);
-                    } else {
-                        nack('API access has been superseded by another frame in this window.');
-                    }
-                } catch (err) {
-                    nack(err);
+                // singleFrameOnly check first so to prevent frame superceding when disabled.
+                if (!data.singleFrameOnly === false || e.sender.isValidWithFrameConnect(e.frameRoutingId)) {
+                    Promise.resolve()
+                        .then(() => action(identity, data, ack, nack))
+                        .then(result => {
+                            if (result !== undefined) {
+                                ack(new AckPayload(result));
+                            }
+                        }).catch(err => {
+                            if (err !== undefined) {
```

Comment:
Shouldn't we be `nack`ing even if the `err` is undefined to guard against an empty `reject` call? 

By leaving it up to the `resolve` and `reject` payloads being defined (while I agree that they should be), by not calling the `ack` / `nack` you can potentially leave a renderer hanging in the worst case (a sync call) or at best leave a caller waiting forever (inviting a memory leak)